tests/integration/ha/ha_helpers.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+import logging
+import re
+from subprocess import PIPE, check_output
+
+from pytest_operator.plugin import OpsTest
+
+from integration.helpers import APP_NAME, get_address
+from literals import SECURITY_PROTOCOL_PORTS
+from snap import KafkaSnap
+
+PROCESS = "kafka.Kafka"
+SERVICE_DEFAULT_PATH = "/etc/systemd/system/snap.charmed-kafka.daemon.service"
+
+logger = logging.getLogger(__name__)
+
+
+class ProcessError(Exception):
+    """Raised when a process fails."""
+
+
+class ProcessRunningError(Exception):
+    """Raised when a process is running when it is not expected to be."""
+
+
+async def get_topic_leader(ops_test: OpsTest, topic: str) -> int:
+    """Get the broker with the topic leader.
+
+    Args:
+        ops_test: OpsTest utility class
+        topic: the desired topic to check
+    """
+    bootstrap_server = (
+        await get_address(ops_test=ops_test)
+        + f":{SECURITY_PROTOCOL_PORTS['SASL_PLAINTEXT'].client}"
+    )
+
+    result = check_output(
+        f"JUJU_MODEL={ops_test.model_full_name} juju ssh kafka/0 sudo -i 'charmed-kafka.topics --bootstrap-server {bootstrap_server} --command-config {KafkaSnap.CONF_PATH}/client.properties --describe --topic {topic}'",
+        stderr=PIPE,
+        shell=True,
+        universal_newlines=True,
+    )
+
+    return re.search(r"Leader: (\d+)", result)[1]
+
+
+async def get_topic_offsets(ops_test: OpsTest, topic: str, unit_name: str) -> list[str]:
+    """Get the offsets of a topic on a unit.
+
+    Args:
+        ops_test: OpsTest utility class
+        topic: the desired topic to check
+        unit_name: unit to check the offsets on
+    """
+    bootstrap_server = (
+        await get_address(ops_test=ops_test)
+        + f":{SECURITY_PROTOCOL_PORTS['SASL_PLAINTEXT'].client}"
+    )
+
+    # example of topic offset output: 'test-topic:0:10'
+    result = check_output(
+        f"JUJU_MODEL={ops_test.model_full_name} juju ssh {unit_name} sudo -i 'charmed-kafka.get-offsets --bootstrap-server {bootstrap_server} --command-config {KafkaSnap.CONF_PATH}/client.properties --topic {topic}'",
+        stderr=PIPE,
+        shell=True,
+        universal_newlines=True,
+    )
+
+    return re.search(rf"{topic}:(\d+:\d+)", result)[1].split(":")
```

Comment:
**question** I'm not sure I fully following the code, but would this code also work in the case of multiple partitions? E.g. 
```
test-topic:0:10
test-topic:1:12
```

If we are not handling this properly, could we either raise an exception (if partitoins is larger than 1) or if it is not too much work, just extend the support for multi-partition topics? I can just see the case that someone else would extend the tests, use this function for multi-partition use-case and get strange results

Reply:
At the moment I didn't add multi partitioned topics on the tests. I think this could be a future extension when we add automatic partition rebalancing.

Reply:
Since this is on tests, I would rather leave this like this. If output were to change, or we add partitioning, I think is best to let the code raise here, so it's obvious that this part also needs to change.